packages/node/src/indexer/api.service.ts

```diff
@@ -89,82 +94,134 @@ export class ApiService implements OnApplicationShutdown {
   }
 
-  async getPatchedApi(
-    blockHash: string | BlockHash,
-    blockNumber: number,
-    parentBlockHash?: string | BlockHash,
-  ): Promise<ApiAt> {
-    this.currentBlockHash = blockHash.toString();
-    this.currentBlockNumber = blockNumber;
-    if (parentBlockHash) {
-      this.currentRuntimeVersion = await this.api.rpc.state.getRuntimeVersion(
-        parentBlockHash,
-      );
+  async getSafeApi(height: number): Promise<CosmosSafeClient> {
+    const { network } = this.project;
+    const client = await CosmosSafeClient.safeConnect(network.endpoint, height);
+    return client;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async getChainType(
+    ds: SubqlProjectDs,
+  ): Promise<Record<string, GeneratedType>> {
+    if (!ds.chainTypes) {
+      return {};
     }
-    const apiAt = (await this.api.at(
-      blockHash,
-      this.currentRuntimeVersion,
-    )) as ApiAt;
-    this.patchApiRpc(this.api, apiAt);
-    return apiAt;
+
+    const res: Record<string, GeneratedType> = {};
+    for (const packages of ds.chainTypes) {
+      const packageName = packages[0];
+      const file = packages[1].file;
+      const messages = packages[1].messages;
+      load(path.join(this.project.root, file), function (err, root) {
+        if (err) throw err;
+        for (const msg of messages) {
+          const msgObj = root.lookupType(`${packageName}.${msg}`);
+          res[`/${packageName}.${msg}`] = msgObj;
+        }
+      });
+    }
+    return res;
   }
+}
 
-  private redecorateRpcFunction<T extends 'promise' | 'rxjs'>(
-    original: RpcMethodResult<T, AnyFunction>,
-  ): RpcMethodResult<T, AnyFunction> {
-    const methodName = this.getRPCFunctionName(original);
-    if (original.meta.params) {
-      const hashIndex = original.meta.params.findIndex(
-        ({ isHistoric }) => isHistoric,
-      );
-      if (hashIndex > -1) {
-        const isBlockNumber =
-          original.meta.params[hashIndex].type === 'BlockNumber';
-
-        const ret = ((...args: any[]) => {
-          const argsClone = [...args];
-          argsClone[hashIndex] = isBlockNumber
-            ? this.currentBlockNumber
-            : this.currentBlockHash;
-          return original(...argsClone);
-        }) as RpcMethodResult<T, AnyFunction>;
-        ret.raw = NOT_SUPPORT(`${methodName}.raw`);
-        ret.meta = original.meta;
-        return ret;
+export class CosmosClient {
+  constructor(
+    private readonly baseApi: CosmWasmClient,
+    private registry: Registry,
+  ) {}
+
+  async chainId(): Promise<string> {
+    return this.baseApi.getChainId();
+  }
+
+  async finalisedHeight(): Promise<number> {
+    return this.baseApi.getHeight();
+  }
+
+  async blockInfo(height?: number): Promise<Block> {
+    return this.baseApi.getBlock(height);
+  }
+
+  async txInfoByHeight(height: number): Promise<readonly IndexedTx[]> {
+    return this.baseApi.searchTx({ height: height });
+  }
+
+  decodeMsg(msg: any) {
```

Comment:
Should this be done for all the message types that are added?

Reply:
yeah, all the messages must be decoded here. But, only the messages types that are in the registry can be decoded without error